canonicalwebteam/discourse_docs/parsers.py

```diff
@@ -0,0 +1,288 @@
+    - The body HTML
+    - The navigation markup
+
+    Set all as properties on the object
+    """
+
+    index = parse_topic(topic)
+    index_soup = BeautifulSoup(index["body_html"], features="html.parser")
+
+    # Get the nav
+    index["body_html"] = str(
+        get_preamble(index_soup, break_on_title="Content")
+    )
+
+    # Parse navigation
+    index["navigation"] = parse_navigation(index_soup)
+
+    return index
+
+
+def parse_topic(topic):
+    """
+    Parse a topic object from the Discourse API
+    and return document data:
+    - title: The title
+    - body_html: The HTML content of the initial topic post
+                    (with some post-processing)
+    - updated: A human-readable date, relative to now
+                (e.g. "3 days ago")
+    - forum_link: The link to the original forum post
+    """
+
+    updated_datetime = dateutil.parser.parse(
+        topic["post_stream"]["posts"][0]["updated_at"]
+    )
+
+    return {
+        "title": topic["title"],
+        "body_html": process_topic_html(
+            topic["post_stream"]["posts"][0]["cooked"]
+        ),
+        "updated": humanize.naturaltime(updated_datetime.replace(tzinfo=None)),
+        "topic_path": f"/t/{topic['slug']}/{topic['id']}",
+    }
+
+
+def parse_navigation(index_soup):
+    """
+    Given the HTML soup of a index topic
+    extract the "navigation" section
+    """
+
+    nav_soup = get_section(index_soup, "Content")
+    nav_html = "Navigation missing"
+
+    if nav_soup:
+        nav_html = str(nav_soup)
+
+    return nav_html
+
+
+def process_topic_html(html):
+    """
+    Given topic HTML, apply post-process steps
+    """
+
+    soup = BeautifulSoup(html, features="html.parser")
+    soup = replace_notifications(soup)
+    soup = replace_notes_to_editors(soup)
+
+    return str(soup)
+
+
+def replace_notes_to_editors(soup):
+    """
+    Given HTML soup, remove 'NOTE TO EDITORS' sections.
+
+    We expect these sections to be of the HTML format:
+
+    <aside class="quote no-group">
+      <blockquote>
+        <p>
+          <img title=":construction:" class="emoji" ...>
+          <strong>NOTE TO EDITORS</strong>
+          <img title=":construction:" class="emoji" ...>
+        </p>
+        <p> ... </p>
+      </blockquote>
+    </aside>
+    """
+
+    notes_to_editors_text = soup.find_all(text="NOTE TO EDITORS")
+
+    for text in notes_to_editors_text:
+        # If this section is of the expected HTML format,
+        # we should find the <aside> container 4 levels up from
+        # the "NOTE TO EDITORS" text
+        container = text.parent.parent.parent.parent
+
+        if container.name == "aside" and "quote" in container.attrs["class"]:
+            container.decompose()
+
+    return soup
+
+
+def replace_notifications(soup):
+    """
+    Given some BeautifulSoup of a document,
+    replace blockquotes with the appropriate notification markup
+
+    E.g.:
+
+        <blockquote><p>ⓘ Content</p></blockquote>
```

Comment:
Same, maybe we can link to the source here.

Reply:
I've tried to clarify.